formats/protobuf/src/main/java/io/cloudevents/protobuf/ProtoDeserializer.java

```diff
@@ -0,0 +1,128 @@
+import io.cloudevents.rw.CloudEventReader;
+import io.cloudevents.rw.CloudEventWriter;
+import io.cloudevents.rw.CloudEventWriterFactory;
+import io.cloudevents.v1.proto.CloudEvent;
+import io.cloudevents.v1.proto.CloudEvent.CloudEventAttributeValue;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.util.Base64;
+import java.util.Map.Entry;
+
+/**
+ * Implements a {@link CloudEventReader} that can deserialize a {@link CloudEvent} protobuf representation;
+ */
+class ProtoDeserializer implements CloudEventReader {
+
+    private static final Base64.Encoder base64Encoder = Base64.getEncoder();
+
+    private final CloudEvent protoCe;
+
+    public ProtoDeserializer(CloudEvent protoCe) {
+        this.protoCe = protoCe;
+    }
+
+    @Override
+    public <W extends CloudEventWriter<R>, R> R read(
+        CloudEventWriterFactory<W, R> writerFactory,
+        CloudEventDataMapper<? extends CloudEventData> mapper) throws CloudEventRWException {
+        SpecVersion specVersion = SpecVersion.parse(this.protoCe.getSpecVersion());
+
+        final CloudEventWriter<R> writer = writerFactory.create(specVersion);
+
+        // Required attributes
+        writer.withContextAttribute(CloudEventV1.ID, this.protoCe.getId());
+        writer.withContextAttribute(CloudEventV1.SOURCE, this.protoCe.getSource());
+        writer.withContextAttribute(CloudEventV1.TYPE, this.protoCe.getType());
+
+        // Optional attributes
+        for (Entry<String, CloudEventAttributeValue> entry : this.protoCe.getAttributesMap().entrySet()) {
+            String name = entry.getKey();
+            CloudEventAttributeValue val = entry.getValue();
+            switch (val.getAttrCase()) {
+                case CE_BOOLEAN:
+                    writer.withContextAttribute(name, val.getCeBoolean());
+                    break;
+                case CE_INTEGER:
+                    writer.withContextAttribute(name, val.getCeInteger());
+                    break;
+                case CE_STRING:
+                    writer.withContextAttribute(name, val.getCeString());
+                    break;
+                case CE_BYTES:
+                    writer.withContextAttribute(name, val.getCeBytes().toByteArray());
+                    break;
+                case CE_URI:
+                    writer.withContextAttribute(name, URI.create(val.getCeUri()));
+                    break;
+                case CE_URI_REF:
+                    writer.withContextAttribute(name, URI.create(val.getCeUriRef()));
+                    break;
+                case CE_TIMESTAMP:
+                    com.google.protobuf.Timestamp timestamp = val.getCeTimestamp();
+                    Instant instant = Instant.ofEpochSecond(timestamp.getSeconds(), timestamp.getNanos());
+                    writer.withContextAttribute(name, instant.atOffset(ZoneOffset.UTC));
+                    break;
```

Comment:
IMO the conversion proto timestamp to `OffsetDateTime` deserves a separate private static method

Reply:
Done